### PR TITLE
Change how we get executable path

### PR DESCRIPTION
**Pull request by MikeHolman**
Right now, it's required that main calls setExecPath or cruntime be located in the CMAKE install directory, in order to find it when linking models. This was causing issues with CompilerLibTest, which can't currently call setExecPath, and so models will fail to link if the environment is not exactly right.

My fix is to remove the need for this call, by calling getMainExecutable where it is needed. The downside to doing it this way is that at the later point, we don't have argv[0], which is used as a fallback mechanism to find the executable on Linux in case /proc is not mounted. This seems like a rare scenario that we don't need to support, and removing it has the benefit of reducing the initialization that users need to do.

### CODE REVIEW

**Comment by MikeHolman** [comment]
If you think this is too risky I am happy to go about the fix in a different way, but I thought this would make things cleaner.

**Comment by sstamenova** [comment]
I think it is a good idea to have a better way to determine where things are, however, I think this change is incomplete in achieving that. A couple of things that would be good to address are:
1) CMAKE_INSTALL_PREFIX (or at least its documentation for onnx-mlir). Right now, if I recall correctly, onnx-mlir has to be installed in the same place as llvm-project for some of the detection to work correctly. If this is no longer the case, then at the very least the comments need to be updated to reflect that. Ideally, we would move to a place where they can be installed to different places.
2) Some of the libraries (such as cruntime) have assumed locations based on where the executable is - whether that is onnx-mlir.exe or one of the test executables.
3) The test executables are placed in the bin directory because of 2). It would be good to remove the need for the test binaries to be bin-placed.

**Comment by MikeHolman** [comment]
I feel that is kind of beyond the scope of this change. I'm just trying to avoid having to resolve the current executable path from main, not trying to rearchitect where/how dependencies are located.

Right now there is a search path where some of the locations are relative to the directory of the executable, which seems sensible to me. This search path for libs can be overridden with the `ONNX_MLIR_RUNTIME_DIR` environment variable, so if you have a particular location you want to use, you can do that as well. This system might not be perfect but it seems good enough.

The only enhancement I would want to make would be to add an environment variable to allow overriding the tool path as well. It seems somewhat tangential to this change, but I can do it if you think it would be useful.

**Comment by gongsu832** [comment]
@sstamenova is right. Currently there are a lot of logic in `CompilerUtils.cpp` depending on `kExecPath` being set to the main process executable. Read the comments above `CompilerUtils.cpp:getToolPath()`.

`getMainExecutable()` must be called on the `main` function to get the correct main executable name, which is why `setExecPath()` must be called from `main()`. The way you changed it, you would get the `libCompilerUtils.{so|dll}` name, not the main executable name, unless `getMainExecutable()` has changed since we wrote the code.

**Comment by MikeHolman** [comment]
It's certainly possible that getMainExecutable has changed, since the code references falling back to the "old" way. On Mac and Windows, neither argument is ever used and OS primitives are always used to get the executable path. On Linux, BSD, Solaris, and some 13 other platforms mainAddr is never used and argv0 is only used in case /proc is not mounted. mainAddr is only used in case of an unknown platform that supports the dladdr function, so it seems plausible that mainAddr is never used on platforms onnx-mlir supports, in which case we can pass nullptr for mainAddr as well.

**Comment by caoimhinuibrian** [comment]
@jenkins-droid test this please

**Comment by gongsu832** [comment]
It's certainly nice not having to call `getMainExecutable()` from `main()`. I remember the itch having to stick `setExecPath()` in the `main()` of a user program. So it's great to be able to get rid of `setExecPath()`. Although I would prefer to keep the `kExecPath` as a fallback (and perhaps produce a warning message) instead of throwing an assertion error.
```
diff --git a/src/Compiler/CompilerUtils.cpp b/src/Compiler/CompilerUtils.cpp
index d377b53..d286427 100644
--- a/src/Compiler/CompilerUtils.cpp
+++ b/src/Compiler/CompilerUtils.cpp
@@ -130,8 +130,7 @@ string getExecPath() {
   // where /proc isn't mounted
   auto execPath =
       llvm::sys::fs::getMainExecutable(nullptr, (void *)getExecPath);
-  assert(!execPath.empty() && "Could not find path to current executable");
-  return execPath;
+  return execPath.empty() ? kExecPath : execPath;
 }

 // Runtime directory contains all the libraries, jars, etc. that are
diff --git a/src/Compiler/ExternalUtil.hpp.in b/src/Compiler/ExternalUtil.hpp.in
index 8784020..6d65f30 100644
--- a/src/Compiler/ExternalUtil.hpp.in
+++ b/src/Compiler/ExternalUtil.hpp.in
@@ -3,6 +3,7 @@
 #include <string>

 namespace onnx_mlir {
+const std::string kExecPath = "@CMAKE_INSTALL_PREFIX@/bin/$<TARGET_FILE_NAME:onnx-mlir>"; /* fallback if not set by main */
 const std::string kInstPath = "@CMAKE_INSTALL_PREFIX@";
 const std::string kOptPath = "$<TARGET_FILE:opt>";
 const std::string kLlcPath = "$<TARGET_FILE:llc>";
```

File: src/Compiler/CompilerUtils.cpp [hunk 1]
```diff
@@ -124,6 +124,17 @@ static bool keepFiles(KeepFilesOfType preserve) {
   return false;
 }
 
+string getExecPath() {
+  // argv0 is only used as a fallback for rare environments
+  // where /proc isn't mounted and mainExecAddr is only needed for
+  // unknown unix-like platforms
+  auto execPath = llvm::sys::fs::getMainExecutable(nullptr, nullptr);
+  std::cerr << "Warning: Could not find path to current executable, falling "
+               "back to default install path."
```
Comment (gongsu832) [inline src/Compiler/CompilerUtils.cpp:133]: Nit: maybe print out kExecPath also?

**Review by gongsu832 — APPROVED** [review]
LGTM

**Comment by AlexandreEichenberger** [comment]
@MikeHolman not sure if you can squash and merge. Let me know if you have more things to push in, or it is ready do land.

**Pull request merged**